cng/aes.go

```diff
@@ -119,34 +119,45 @@ func (c *aesCipher) NewGCMTLS() (cipher.AEAD, error) {
 	return newGCM(c.key, true)
 }
 
-type aesCBC struct {
-	kh      bcrypt.KEY_HANDLE
-	iv      [aesBlockSize]byte
-	encrypt bool
+type cbcCipher struct {
+	kh bcrypt.KEY_HANDLE
+	// us aesBlockSize as the max block size for all ciphers
+	// to avoid allocations.
+	iv        [aesBlockSize]byte
+	blockSize int
+	encrypt   bool
 }
 
-func newCBC(encrypt bool, key, iv []byte) *aesCBC {
-	kh, err := newCipherHandle(bcrypt.AES_ALGORITHM, bcrypt.CHAIN_MODE_CBC, key)
+func newCBC(encrypt bool, alg string, key, iv []byte) *cbcCipher {
+	kh, err := newCipherHandle(alg, bcrypt.CHAIN_MODE_CBC, key)
 	if err != nil {
 		panic(err)
 	}
-	x := &aesCBC{kh: kh, encrypt: encrypt}
+	x := &cbcCipher{kh: kh, encrypt: encrypt}
+	switch alg {
+	case bcrypt.AES_ALGORITHM:
+		x.blockSize = aesBlockSize
+	case bcrypt.DES_ALGORITHM:
+		x.blockSize = desBlockSize
+	default:
+		panic("invalid algorithm: " + alg)
+	}
 	x.SetIV(iv)
-	runtime.SetFinalizer(x, (*aesCBC).finalize)
+	runtime.SetFinalizer(x, (*cbcCipher).finalize)
```

Comment:
Maybe move this just below `x := &cbcCipher{kh: kh, encrypt: encrypt}` to avoid leaking if a panic happens (and is handled)?

Reply:
Done!